src/lib/db/access-store.ts

```diff
@@ -53,75 +68,139 @@ export class AccessStore implements IAccessStore {
+            .join(`${T.PERMISSIONS} AS p`, 'p.id', 'rp.permission_id')
+            .where('ur.user_id', '=', userId);
         stopTimer();
-        return rows;
+        return rows.map(this.mapUserPermission);
     }
 
-    async getRoles(): Promise<IRole[]> {
-        return this.db
-            .select(['id', 'name', 'type', 'description'])
-            .from<IRole>(T.ROLES);
-    }
+    mapUserPermission(row: IPermissionRow): IUserPermission {
+        let project;
+        // Since the editor should have access to the default project,
+        // we map the project to the project and environment specific
+        // permissions that are connected to the editor role.
+        if (row.role_id === EDITOR_ID && row.type !== 'root') {
+            project = 'default';
+        } else if (row.type !== 'root') {
+            project = row.project ? row.project : undefined;
+        }
 
-    async getRoleWithId(id: number): Promise<IRole> {
-        return this.db
-            .select(['id', 'name', 'type', 'description'])
-            .where('id', id)
-            .first()
-            .from<IRole>(T.ROLES);
+        const environment =
+            row.type === 'environment' ? row.environment : undefined;
+        return {
+            project,
+            environment,
+            permission: row.permission,
+        };
     }
 
-    async getRolesForProject(projectId: string): Promise<IRole[]> {
-        return this.db
-            .select(['id', 'name', 'type', 'project', 'description'])
-            .from<IRole>(T.ROLES)
-            .where('project', projectId)
-            .andWhere('type', 'project');
+    async getPermissionsForRole(roleId: number): Promise<IPermission[]> {
+        const stopTimer = this.timer('getPermissionsForRole');
+        const rows = await this.db
+            .select(
+                'p.id',
+                'p.permission',
+                'rp.environment',
+                'p.display_name',
+                'p.type',
+            )
+            .from<IPermission>(`${T.ROLE_PERMISSION} as rp`)
+            .join(`${T.PERMISSIONS} as p`, 'p.id', 'rp.permission_id')
+            .where('rp.role_id', '=', roleId);
+        stopTimer();
+        return rows.map((permission) => {
+            return {
+                id: permission.id,
+                name: permission.permission,
+                environment: permission.environment,
+                displayName: permission.display_name,
+                type: permission.type,
+            };
+        });
     }
 
-    async getRootRoles(): Promise<IRole[]> {
-        return this.db
-            .select(['id', 'name', 'type', 'project', 'description'])
-            .from<IRole>(T.ROLES)
-            .andWhere('type', 'root');
+    async addEnvironmentPermissionsToRole(
+        role_id: number,
+        permissions: IPermission[],
+    ): Promise<void> {
+        const rows = permissions.map((x) => {
```

Comment:
This is a discussion to be had I guess, but I personally can't stand using placeholder arguments that bears no meaning such as 'x'. My main objection is the added cognitive load it takes to figure out what `x` is when something like `permission` would allow you to automatically associate it with the data structure for permission.

I think that as long as we have bundlers such as webpack that handles minifying and optimizing our code for us, we should strive to keep the code as readable and easy to maintain as possible.

Reply:
💯 agree with @FredrikOseberg 👍🏼

Reply:
Yeah I think your argument is a great one and I don't think there's an actual reason not to do this. There are some things that I think we should consider here though:

1) This is common enough in map/reduce/filter/loop/one line lambda that if we want this to be the case going forward we should decide this as a team and codify it into a document of code standards/guidelines. I think linters can only take you so far without also understanding why the code should be the way it is (I'm tempted to just copy paste this comment). And I think we should do this soon, before we hire more devs and we start to get pockets of dissenting opinions throughput the company. I think you guys are 100% correct here and I don't think anyone else on the team would disagree so we should do this now(ish)

2) The code is project in its current state is full of one character (or short enough not to be meaningfully different) variable names where map/filter is concerned. If we're going to shift towards a more rational naming standard we should make it a priority to also change this in the project, I think there's almost as much value in consistent standards as there are good ones.

I've updated the 3 places where x as a variable name has been pointed out (be the change you want and all that)

Reply:
I think that is in excellent point, and we should definitely standardize some of this. I actually think this is a fantastic resource that should be part of our team members curriculum: https://github.com/ryanmcdermott/clean-code-javascript

It's made for js, but the principles apply everywhere.

Reply:
That is a seriously nice link. I think we should pursue this conversation further. Docs for newbies would be amazing as we scale and consistent code would be rad. Going to close this one for one